src/adapters/ibm-vllm/utils/build-client.ts

```diff
@@ -0,0 +1,140 @@
+  "grpc.max_receive_message_length": 32 * 1024 * 1024, // 32MiB
+} as const;
+
+interface BuildClientProps {
+  cache?: LRUCache<string, GenerationServiceClient>;
+  modelRouterSubdomain?: string;
+  url?: string;
+  credentials?: {
+    rootCert: string;
+    certChain: string;
+    privateKey: string;
+  };
+  clientShutdownTimeout?: number;
+}
+
+const generationPackageObject = grpc.loadPackageDefinition(
+  protoLoader.loadSync([GENERATION_PROTO_PATH.pathname], {
+    longs: Number,
+    enums: String,
+    arrays: true,
+    objects: true,
+    oneofs: true,
+    keepCase: true,
+    defaults: true,
+  }),
+) as unknown as GenerationProtoGentypes;
+
+const defaultClientCache = new LRUCache<string, GenerationServiceClient>({
+  max: 100,
+  ttl: DEFAULT_GRPC_CLIENT_TTL,
+});
+
+export function buildClient({
+  url = getEnv("IBM_VLLM_URL"),
+  credentials = {
+    rootCert: getEnv("IBM_VLLM_ROOT_CERT"),
+    privateKey: getEnv("IBM_VLLM_PRIVATE_KEY"),
+    certChain: getEnv("IBM_VLLM_CERT_CHAIN"),
+  },
+  cache = defaultClientCache,
+  clientShutdownTimeout = DEFAULT_GRPC_CLIENT_SHUTDOWN_TIMEOUT,
+}: BuildClientProps): GenerationServiceClient {
+  if (!credentials.rootCert || !credentials.privateKey || !credentials.certChain) {
+    throw new ValueError(
+      [
+        `IBM vllm GRPC credentials were not provided. Either set them directly or put them in ENV:`,
+        `"IBM_VLLM_ROOT_CERT", "IBM_VLLM_PRIVATE_KEY", "IBM_VLLM_CERT_CHAIN"`,
+      ].join("\n"),
+    );
+  }
+  if (!url) {
+    throw new ValueError(
+      `IBM vllm URL was not provided. Either set them directly or put them in ENV: "IBM_VLLM_URL"`,
+    );
+  }
+
+  return new Proxy(
+    {},
+    {
+      get: (_target, prop: keyof GenerationServiceClient) => {
+        return function <
+          K extends keyof GenerationServiceClient,
+          P extends GenerationServiceClient[K],
+        >(...params: Parameters<P>) {
+          const firstParam = params?.[0] as Exclude<(typeof params)[0], string | Deadline>;
+          const modelId = firstParam?.["model_id"];
+          if (!modelId) {
+            throw new FrameworkError("model_id must be provided in the first GRPC method argument");
+          }
+          const modelSpecificUrl = url.replace(/{model_id}/, modelId.replaceAll("/", "--"));
+          const createClient = () => {
+            const client = new generationPackageObject.fmaas.GenerationService(
+              url,
+              grpc.credentials.createSsl(
+                Buffer.from(credentials.rootCert),
+                Buffer.from(credentials.privateKey),
+                Buffer.from(credentials.certChain),
+              ),
+              options,
+            );
+            setTimeout(() => {
+              try {
+                cache.delete(modelSpecificUrl);
+                client.close();
+              } catch {
+                /* empty */
+              }
+            }, clientShutdownTimeout).unref();
```

Comment:
This is the cleanup/caching mechanism for memory leaks ported from API, but with a few differences:
- Client is a method `Proxy`. Before, `buildClient` returned a cached version of client which destroys itself in a few minutes, this is bad if you create a global instance in your code. With the Proxy, you get a new or cached client everytime you call a method on it
- `unref()` on `setTimeout`, because otherwise it holds up the node process with an unresolved Promise
- Using different cache package `lru-cache` - because `cache-manager` we used was updated and is requiring serializable objects now, which GRPC client is not. (Although the LRU might be an overkill, you'd need to call more than 100 different models for it to work, which is impossible), should I refactor this with a `Map`?

Reply:
Using the following mechanism is not needed.

You are ready to go by adding a class-getter with a cache decorator.
If you need to do cleanup, then introduce `destroy` method on it. You can add an auto-destroy mechanism in there if you want to.

See: 
- https://github.com/i-am-bee/bee-agent-framework/blob/main/src/tools/database/sql.ts#L84-L97
- https://github.com/i-am-bee/bee-agent-framework/blob/main/src/tools/database/sql.ts#L146-L160

Reply:
If you would like to have a factory separately, you can use the `CacheFn` construct.